calendar.1
.\"	$MirOS: src/usr.bin/calendar/calendar.1,v 1.27 2025/06/20 17:32:14 tg Exp $
.\"	$OpenBSD: calendar.1,v 1.29 2005/03/22 10:57:27 xsa Exp $
.\"
.\" Copyright (c) 1989, 1990, 1993
.\"     The Regents of the University of California.  All rights reserved.
.\" Copyright © 2019, 2021, 2023, 2025
.\"	mirabilos <m$(date +%Y)@mirbsd.de>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)calendar.1  8.1 (Berkeley) 6/29/93
.\"
.\"-
.\" $miros: contrib/samples/portmdoc,v 1.27 2025/06/20 17:30:44 tg Exp $
.\"-
.\" Copyright © 2008, 2009, 2010, 2016, 2018, 2020, 2023, 2025
.\"	mirabilos <m$(date +%Y)@mirbsd.de>
.\" Copyright © 1991, 1993
.\"	The Regents of the University of California. (Ⓕ 3-clause BSD)
.\" All rights reserved.
.\"
.\" Glue GNU groff (BSD and GNU mdoc both) to AT&T nroff (UCB mdoc).
.\" * ` generates ‘ in gnroff, so use \`
.\" * ' generates ’ in gnroff, \' generates ´, so use \*(aq
.\" * - generates ‐ in gnroff, \- generates −, so .tr it to -
.\"   thus use - for hyphens and \- for minus signs and option dashes
.\" * ~ is size-reduced and placed atop in groff, so use \*(TI
.\" * ^ is size-reduced and placed atop in groff, so use \*(ha
.\" * \(en does not work in nroff, so use \*(en for a solo en dash
.\" *   and \*(EM for a correctly spaced em dash (omit space around!)
.\" * <>| are problematic, so redefine and use \*(Lt\*(Gt\*(Ba
.\" Also make sure to use \& *before* a punctuation char that is to not
.\" be interpreted as punctuation, and especially with two-letter words
.\" but also (after) a period that does not end a sentence (“e.g.\&”).
.\"-
.\"
.\" Implement .Dd with the Mdocdate RCS keyword
.\"
.rn Dd xD
.de Dd
.ie \\$1$Mdocdate: \{\
.	xD \\$2 \\$3, \\$4
.\}
.el .xD \\$1 \\$2 \\$3
..
.\"
.\" .Dd must come before most everything, because when called
.\" with -mandoc it loads -mdoc via .so in .Dd (first macro).
.\"
.Dd $Mdocdate: June 20 2025 $
.\"
.\" Check which macro package we use, and do other -mdoc setup.
.\"
.ie \n(.g \{\
.	if n .ss \n[.ss] 0
.	if \*[.T]ascii .tr \-\N'45'
.	if \*[.T]latin1 .tr \-\N'45'
.	if \*[.T]utf8 .tr \-\N'45'
.	if \*[.T]utf8 .tr \(la\*(Lt
.	if \*[.T]utf8 .tr \(ra\*(Gt
.	ds <= \(<=
.	ds >= \(>=
.	ds Rq \(rq
.	ds Lq \(lq
.	ds sL \(aq
.	ds sR \(aq
.	if \*[.T]utf8 .ds sL `
.	if \*[.T]ps .ds sL `
.	if \*[.T]utf8 .ds sR '
.	if \*[.T]ps .ds sR '
.	ds aq \(aq
.	ds TI \(ti
.	ds ha \(ha
.	ds en \(en
.	ie d volume-ds-1 .ds tT gnu
.	el .ie d doc-volume-ds-1 .ds tT gnp
.	el .ds tT bsd
.\}
.el \{\
.	ds aq '
.	ds TI ~
.	ds ha ^
.	ds en \(em
.	ds tT ucb
.\}
.ie n \{\
.	ds EM \ \(em\ \&
.\}
.el \{\
.	ds EM \f(TR\|\(em\|\fP\&
.\}
.\"
.\" Add UCB mdoc compatibility to GNU mdoc
.\" Implement .Mx (MirBSD)
.\"
.ie "\*(tT"gnu" \{\
.	ds sP \s0
.	ds tN \*[Tn-font-size]
.	eo
.	de Mx
.	nr curr-font \n[.f]
.	nr curr-size \n[.ps]
.	ds str-Mx \f[\n[curr-font]]\s[\n[curr-size]u]
.	ds str-Mx1 \*(tN\%MirBSD\*[str-Mx]
.	if !\n[arg-limit] \
.	if \n[.$] \{\
.	ds macro-name Mx
.	parse-args \$@
.	\}
.	if (\n[arg-limit] > \n[arg-ptr]) \{\
.	nr arg-ptr +1
.	ie (\n[type\n[arg-ptr]] == 2) \
.	ds str-Mx1 \*(tN\%MirBSD\~#\*[arg\n[arg-ptr]]\*[str-Mx]
.	el \
.	nr arg-ptr -1
.	\}
.	ds arg\n[arg-ptr] "\*[str-Mx1]
.	nr type\n[arg-ptr] 2
.	ds space\n[arg-ptr] "\*[space]
.	nr num-args (\n[arg-limit] - \n[arg-ptr])
.	nr arg-limit \n[arg-ptr]
.	if \n[num-args] \
.	parse-space-vector
.	print-recursive
..
.	de Aq
.	if !\n[arg-limit] \
.	ds macro-name Aq
.	ie \n[in-authors-section] \{\
.		ds quote-left \*(Lt
.		ds quote-right \*(Gt
.	\}
.	el \{\
.		ds quote-left \[la]
.		ds quote-right \[ra]
.	\}
.	enclose-string \$@
..
.	ec
.\}
.el .ie "\*(tT"gnp" \{\
.	ds sP \s0
.	ie t .ds tN \s[(\n[.ps]u-1z)]
.	el .ds tN
.	eo
.	de Mx
.	nr doc-curr-font \n[.f]
.	nr doc-curr-size \n[.ps]
.	ds doc-str-Mx \f[\n[doc-curr-font]]\s[\n[doc-curr-size]u]
.	ds doc-str-Mx1 \*(tN\%MirBSD\*[doc-str-Mx]
.	if !\n[doc-arg-limit] \
.	if \n[.$] \{\
.	ds doc-macro-name Mx
.	doc-parse-args \$@
.	\}
.	if (\n[doc-arg-limit] > \n[doc-arg-ptr]) \{\
.	nr doc-arg-ptr +1
.	ie (\n[doc-type\n[doc-arg-ptr]] == 2) \
.	ds doc-str-Mx1 \*(tN\%MirBSD\~#\*[doc-arg\n[doc-arg-ptr]]\*[doc-str-Mx]
.	el \
.	nr doc-arg-ptr -1
.	\}
.	ds doc-arg\n[doc-arg-ptr] "\*[doc-str-Mx1]
.	nr doc-type\n[doc-arg-ptr] 2
.	ds doc-space\n[doc-arg-ptr] "\*[doc-space]
.	nr doc-num-args (\n[doc-arg-limit] - \n[doc-arg-ptr])
.	nr doc-arg-limit \n[doc-arg-ptr]
.	if \n[doc-num-args] \
.	doc-parse-space-vector
.	doc-print-recursive
..
.	am1 Sh
.	if d doc-sec-head .if !"\*[doc-sec-head]"" \{\
.		if "\*[doc-sec-head]"\*[doc-section-synopsis]" .na
.		if "\*[doc-sec-head]"\*[doc-section-see-also]" .na
.	\}
..
.	ec
.\}
.el \{\
.	de Mx
.	nr cF \\n(.f
.	nr cZ \\n(.s
.	ds aa \&\f\\n(cF\s\\n(cZ
.	if \\n(aC==0 \{\
.		ie \\n(.$==0 \&\\*(tNMirBSD\\*(aa
.		el .aV \\$1 \\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.	\}
.	if \\n(aC>\\n(aP \{\
.		nr aP \\n(aP+1
.		ie \\n(C\\n(aP==2 \{\
.			nr xX 0
.			nr xX 1+\\*(A\\n(aP
.			as b1 \&\\*(tNMirBSD\ \&
.			if \\n(xX>0 .as b1 #\&
.			as b1 \&\\*(A\\n(aP\\*(aa
.			rr xX
.			ie \\n(aC>\\n(aP \{\
.				nr aP \\n(aP+1
.				nR
.			\}
.			el .aZ
.		\}
.		el \{\
.			as b1 \&\\*(tNMirBSD\\*(aa
.			nR
.		\}
.	\}
..
.\}
.\" </MirCVS://contrib/samples/portmdoc>
.\"-
.Dt CALENDAR 1
.Os
.Sh NAME
.Nm calendar
.Nd reminder service
.Sh SYNOPSIS
.Bk -words
.Nm calendar
.Op Fl abP
.Op Fl A Ar num
.Op Fl B Ar num
.Op Fl D Ar macro Ns Op Ar =defn
.Op Fl f Ar calendarfile
.Op Fl I Ar dir
.Op Fl t Ar [[[cc]yy]mm]dd
.Ek
.Sh DESCRIPTION
The
.Nm
utility checks the current directory or the directory specified by the
.Ev CALENDAR_DIR
environment variable for a file named
.Pa calendar
and displays lines that begin with either today's date
or tomorrow's.
On Fridays, events on Friday through Monday are displayed.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A Ar num
Print lines from today and next
.Ar num
days (forward, future).
.It Fl a
Process the
.Pa calendar
files in all users' home directories and mail the results to them.
This requires superuser privileges.
.It Fl B Ar num
Print lines from today and previous
.Ar num
days (backward, past).
.It Fl b
Enforce special date calculation mode for Cyrillic calendars.
.It Fl D Ar macro Ns Op Ar =defn
Passed on to
.Xr cpp 1 ,
for
.Li #ifdef
etc.\& in the calendar file.
.It Fl f Ar calendarfile
Use
.Ar calendarfile
.Pq So Li \-\& Sc for stdin
as the default calendar file.
.It Fl I Ar dir
Passed on to
.Xr cpp 1 ,
for
.Li #include
in the calendar file.
.It Fl P
Parse the entire calendar file and convert it to an output format
intended to be consumed by calendar format conversion utilities,
representing the information from the original entries including
year and time-of-day extensions (see below).
When given twice, the original first line is also quoted before
each event heading.
.It Fl t Ar [[[cc]yy]mm]dd
Act like the specified value is
.Dq today ,
instead of using the current date.
Two-digit years map to 1969 through 2068.
.El
.Pp
To handle calendars in different encodings, specify
.Dq LANG=<locale_name>
in the calendar file as early as possible.
That being said, there are no locales on
.Mx
so
.Nm
handles
.Ar locale_name
as
.Ar country Ns Li . Ns Ar encoding ;
.Ar country
is used to make some special handling available, and
.Ar encoding
is used to convert the entries to UTF-8.
.Pp
National names for special days can be handled by adding lines such as
.Dq Advent=<national_name>
(for the Western Church first Sunday of Advent),
.Dq Easter=<national_name>
(for Catholic Easter),
.Dq Paskha=<national_name>
(for Orthodox Easter) or
.Dq Pesach=<national_name>
(for Jewish Passover),
setting the corresponding variables.
.Pp
The
.Dq CALENDAR
variable can be used to specify the style.
Only
.Sq Julian
and
.Sq Gregorian
styles are currently supported.
Use
.Dq CALENDAR=
to return to the default (Gregorian).
.Pp
To enforce special date calculation mode for Cyrillic calendars
specify
.Dq LANG=<locale_name>
and
.Dq BODUN=<bodun_prefix>
where <locale_name> can be
.Li ru_RU.KOI8\-R ,
.Li uk_UA.KOI8\-U ,
.Li by_BY.KOI8\-B
or their
.Li .UTF\-8
counterparts.
.Pp
To enable specific anniversary handling, applied to all complete
event texts that start with a four-digit year plus comma and space,
or that end with comma and space and a four-digit year,
appending a formatted text instead, set
.Dq ANNIV=<ago_text><us><cur_text><us><future_text>
(where
.Ic us
is the ASCII unit separator (Ctrl-_, 0x1F) and the texts are used
if the event happened in the past, present or future, respectively;
.Dq "%1$d"
and
.Dq "%2$d"
in the texts are replaced by the year and the absolute year delta
respectively),
.Em or
use
.Dq ANNIV=1
which enables this mode using the default
.Dq "%1$d: %2$d year(s) ago",
.Dq "%1$d: this year" ,
.Dq "%1$d: in %2$d year(s)"
if none are specified otherwise.
Some crazy Russian formats are also parsed,
conflating "\-e" ('s) and "G" (year), ignoring "do n.e." (BCE).
.Pp
Other lines should begin with a month and day.
They may be entered in almost any format, either numeric or as character
strings.
If proper locale is set, national months and weekdays
names can be used (not on
.Mx
which does not support locales).
A single asterisk
.Pq Sq *\&
matches every month.
A day without a month matches that day of every week.
A month without a day matches the first of that month.
Two numbers default to the month followed by the day.
Lines with leading tabs default to the last entered date, allowing
multiple line specifications for a single date.
.Ic Advent ,
.Ic Easter ,
.Ic Paskha
and
.Ic Pesach ,
both these standard English and the user-defined national (see above) names,
correspond to the respective special day for this year; they may be followed
by a positive (with an explicit
.Ql +\& )
or negative integer to offset the day.
Weekday names may be followed by
.Dq \-4
.Li ...\&
.Dq +5
(aliases last, first, second, third, fourth) for moving events like
.Dq the last Monday in April .
.Pp
By convention, dates followed by an asterisk
.Pq Sq *\&
are not fixed, i.e.\& change from year to year.
.Pp
Day descriptions start after the first <tab> character in the line;
if the line does not contain a <tab> character, it isn't printed out.
If the first character in the line is a <tab> character, it is treated as
the continuation of the previous description.
.Pp
Day descriptions may begin with a four-digit year (between 1000 and 2999)
followed by a comma and space, and/or a time-of-day descriptor followed
by a space (in either case, at least one other character is required).
This is mostly useful in
.Fl P
.Pq conversion
mode, but the leading year is picked up by anniversary processing
.Pq see above
as well, and the user may appreciate this information anyway.
A time-of-day specification shall be
.Xo
.Ar starttime Ns Oo Ic dash Ns Oo
.Ar endtime Oc Oc Ns Oo Ic open-bracket
.Ar timezone Ic close-bracket Oc
.Xc
where
.Ar starttime
and
.Ar endtime
are 24-hour specifications (e.g.\& 01:23),
.Ic dash
is either ASCII
.Ql \-\&
hyphen-minus or U+2013 EN DASH),
.Ic open-bracket No is Ql [\& ,
.Ic close-bracket No is Ql ]\&
and
.Ar timezone
is an Olson tzdata identifier.
24-hour specifications go from 00:00 to 23:59; 24:00 is additionally
recognised but mapped to 23:59 so events stay within one day.
.Pp
The parse-and-extract mode output format is as follows,
suppressing the normal output:
.Bl -tag -width Ds
.It \(dg quoted original line
If
.Fl PP
is used, the original first line of an event specification (the one
that does not start with a tab), prefixed by
.Ql \*(Gt ,
is output first (unchanged but preprocessed).
.It \(dg metadata line (event heading)
This line is comprised of the following fields:
.Bl -tag -width Ds
.It variable-day marker
Either an asterisk
.Ql *\&
(if
.Nm
would otherwise have printed one) or an equals sign
.Ql =\& .
This field, unlike all others, is not followed by a space.
.It date
YYYY\-MM\-DD formatted date of either the first occurrence within
a given year, if a year is part of the extended day description
.Pq see above ,
or the next occurrence after or starting with
.Dq today
.Pq see above .
.It recurrence mode
This is normally one of
.Ic yearly ,
.Ic monthly
or
.Ic weekly .
Yearly events with a given year in the extended day description change to
.Ic once .
Events whose days are based on a special calculation (Advent, Easter, etc.) are
.Ic special
instead, independent on whether they would occur once or yearly.
.It recurrence frequency
Special events show the English name (Advent, Easter, Paskha, Pesach),
followed by either a plus or minus, and a number (offset in days).
Yearly and once events show
.Ic \-mm\-dd
or
.Ic \-mm,www+n
where
.Ar mm
is the month number (01 to 12),
.Ar dd
is the day number (01 to 31),
.Ar www
is the weekday (Sun, Mon, Tue, Wed, Thu, Fri, Sat), and
.Ar +n
is either plus or minus and a number, showing the first (+1),
second (+2), etc.\& or last (\-1), penultimate (\-2), etc.\&
occurrence of the given weekday in a month.
Monthly events show
.Ic dd
if they recur on the same day-of-month number, and
.Ar www+n
if they recur on one given weekday each month.
Weekly events show
.Ic www .
.It year
Either an asterisk
.Pq Ql *\&
or a four-digit year number.
.It start and end times and timezone
This is either one field, literally
.Ic whole-day ,
or two fields, optionally followed by a third.
The first field is a time as
.Ic hh:mm ;
the second field either an
.Dq at
.Ql @\&
.Pq denoting no end time was given
or a time as
.Ic hh:mm .
The third field
.Pq if present
is a timezone name
.Pq see above .
.El
.It \(dg description
One or more lines prefixed by a tab character.
Together, they comprise the day description with the extended part removed.
Line breaks are the same as in the calendar file.
.El
.Pp
.Pa https://github.com/mirabilos/ev\-useful/tree/master/cal2ics
is a sample postprocessor converting this format to iCalendar (*.ics) format.
.Pp
The calendar file is preprocessed by
.Xr cpp 1 ,
allowing the inclusion of shared files such as company holidays or
meetings.
If the shared file is not referenced by a full pathname,
.Xr cpp 1
searches in the current (or home) directory first, and then in the
directory
.Pa /usr/share/calendar .
Empty lines and C comments
.Pq Li /* ...\& */
are ignored.
.Pp
Some possible calendar entries
(a \fB\et\fR sequence denotes a <tab> character):
.Bd -unfilled -offset indent
LANG=C
Easter=Ostern

#include <calendar.usholiday>
#include <calendar.birthday>

6/15\fB\et\fRJune 15 (if ambiguous, will default to month/day).
Jun.\& 15\fB\et\fRJune 15.
15 June\fB\et\fRJune 15.
Thursday\fB\et\fREvery Thursday.
June\fB\et\fREvery June 1st.
15 *\fB\et\fR15th of every month.

May Sun+2\fB\et\fRsecond Sunday in May (Muttertag)
04/SunLast\fB\et\fRlast Sunday in April,
\fB\et\fRsummer time in Europe
Easter\fB\et\fREaster
Ostern\-2\fB\et\fRGood Friday (2 days before Easter)
Paskha\fB\et\fROrthodox Easter
Advent+21\fB\et\fR4th Sunday of Advent
.Ed
.Sh ENVIRONMENT
.Bl -tag -width CALENDAR_DIR -compact
.It Ev CALENDAR_DIR
Unless
.Fl a
is given, this directory instead of the current working directory
is used as base directory to search for
.Pa calendar
files in.
.El
.Sh FILES
.Bl -tag -width ~/.etc/calendar/nomail -compact
.It Pa calendar
File in current directory
.Pq Fl f No overrides .
.It Pa ~/.etc/calendar
In
.Fl a
mode, or when no
.Pa calendar
file exists in
.Ev $CALENDAR_DIR
or the current directory,
.Nm
changes into this directory in the user's
.Ev $HOME
and looks again for the
.Pa calendar No file Pq Fl f .
.It Pa ~/.etc/calendar/nomail
.Nm
.Fl a
will not mail if this file exists.
.It Pa calendar.all
International and national calendar files.
.It Pa calendar.birthday
Births and deaths of famous (and not-so-famous) people.
.It Pa calendar.christian
Christian holidays (should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year).
.It Pa calendar.computer
Computer people's days of special significance.
.It Pa calendar.croatian
Croatian calendar.
.It Pa calendar.fictional

debian/changelog
calendar-mirbsd (20250620) wtf; urgency=high

  * Merge from CVS:
    • update portmdoc to cope with groff’s incompatible changes
      in trixie/sid, correct .An angle brackets, formatting fixes
    • minor data fixes/updates in calendars/
    • fix the way calendar files are searched: -a only searches
      below the ~username/.etc/calendar directory, sanitise others
    • make -t optarg parsing more reliable
    • fixes in the manpage

 -- Thorsten Glaser <tg@mirbsd.de>  Fri, 20 Jun 2025 17:41:14 +0000

calendar-mirbsd (20230603) wtf; urgency=medium

  * Merge from CVS:
    • pass -D/-I flags to cpp(1)

 -- Thorsten Glaser <tg@mirbsd.de>  Sat, 03 Jun 2023 23:54:42 +0000

calendar-mirbsd (20230516) wtf; urgency=medium

  * Merge from CVS:
    • some cleanup of calendar entries, deduplicate, fix typos
    • allow open-ended events (“17:00–”) in the time-of-day spec
    • some cleanup of code and comments
    • issue diagnostics for some parse errors (to catch mistakes)
    • do output empty tab-led continuation lines
  * Update porting code for calendar_bodunratecheck()
  * opt out of buggy GCC LTO
  * Update UMEGAYA upstream IRC server
  * CPPFLAGS: enable LFS for glibc, thanks lintian
  * Depends change cpp to cpp:any for Multi-Arch installability

 -- Thorsten Glaser <tg@mirbsd.de>  Tue, 16 May 2023 20:18:42 +0200

calendar-mirbsd (20211101) wtf; urgency=medium

  * Merge from CVS:
    • another fix to endpwent(3) calling
  * Drop uid/gid/groups when switching users for “calendar -a”; at
    least (Closes: #998206) even though no proper fix (it does not
    load target users’ group vector, environment or whatever else)
  * Fix autopkgtest: fudge the current day so the output is right

 -- Thorsten Glaser <tg@mirbsd.de>  Sat, 06 Nov 2021 20:57:26 +0100

calendar-mirbsd (20211031) wtf; urgency=low

  * Initial package.

 -- Thorsten Glaser <tg@mirbsd.de>  Sun, 31 Oct 2021 23:48:32 +0100


calendar.c
/*	$OpenBSD: calendar.c,v 1.23 2004/12/10 15:31:01 mickey Exp $	*/

/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 * Copyright (c) 2021, 2023, 2025
 *	mirabilos <m$(date +%Y)@mirbsd.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef MBSDPORT_H
#include MBSDPORT_H
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <err.h>
#include <errno.h>
#include <grp.h>
#include <locale.h>
#ifndef USE_CUSTOM_USERSWITCH
#include <login_cap.h>
#endif
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "pathnames.h"
#include "calendar.h"

#ifndef USE_CUSTOM_USERSWITCH
#define userswitch(pw)	setusercontext(NULL, (pw), (pw)->pw_uid, \
			    LOGIN_SETALL ^ LOGIN_SETLOGIN)
#endif

__IDSTRING(pathnames_h, PATHNAMES_H);
__IDSTRING(calendar_h, CALENDAR_H);

__COPYRIGHT("Copyright (c) 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.");
__SCCSID("@(#)calendar.c  8.3 (Berkeley) 3/25/94");
__RCSID("$MirOS: src/usr.bin/calendar/calendar.c,v 1.23 2025/06/20 01:20:38 tg Exp $");

const char *calendarBaseDir = ".";		/* default base directory */
const char *calendarFile = "calendar";		/* default calendar file */
const char calendarHome[] = ".etc/calendar";	/* alternate location below $HOME */
static const char calendarNoMail[] = "nomail";	/* don't sent mail if this file exists */

struct passwd *pw;
unsigned char doall = 0;
unsigned char bodun_always = 0;
unsigned char parsecvt = 0;
time_t f_time = 0;

int f_dayAfter = 0; /* days after current date */
int f_dayBefore = 0; /* days before current date */

struct specialev spev[NUMEV];

char *cppargv[NCPPARGV];

static void childsig(int);
static void addcppargv(const char *, int);

int
main(int argc, char *argv[])
{
	int ch;
	int daysAB = 0;

	(void)setlocale(LC_ALL, "");

	addcppargv("cpp", 2);
	addcppargv("-traditional", 2);
	addcppargv("-undef", 2);
	addcppargv("-U__GNUC__", 2);
#ifdef UNICODE
	addcppargv("-DUNICODE", 2);
#endif
	addcppargv("-P", 2);
	addcppargv("-I.", 2);
	addcppargv(_PATH_INCLUDE, 2);

	while ((ch = getopt(argc, argv, "A:aB:bD:f:I:Pt:-")) != -1)
		switch (ch) {
		case '-':		/* backward contemptible */
		case 'a':
			if (getuid())
				errx(1, "%s", strerror(EPERM));
			doall = 1;
			break;

		case 'b':
			bodun_always = 1;
			break;

		case 'D':
		case 'I':
			addcppargv(optarg, ch);
			break;

		case 'f': /* other calendar file */
			if (!*optarg || *optarg == '/')
				errx(1, "invalid calendar filename");
			calendarFile = optarg;
			break;

		case 'P': /* parse/convert mode */
			++parsecvt;
			break;

		case 't': /* other date, undocumented, for tests */
			if ((f_time = Mktime(optarg)) <= 0)
				errx(1, "specified date is outside allowed range");
			break;

		case 'A': /* days after current date */
			f_dayAfter = atoi(optarg);
			daysAB = 1;
			break;

		case 'B': /* days before current date */
			f_dayBefore = atoi(optarg);
			if (f_dayBefore)
				daysAB = 1;
			break;

		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc)
		usage();

	addcppargv(calendarFile, 1);
	addcppargv(NULL, 0);

	/* use current time */
	if (f_time <= 0)
		(void)time(&f_time);

	if (f_dayBefore) {
		/* Move back in time and only look forwards */
		f_dayAfter += f_dayBefore;
		f_time -= 86400 * f_dayBefore;
		f_dayBefore = 0;
	}
	settime(daysAB);

	if (doall) {
		pid_t kid, deadkid;
		int kidstat, kidreaped, runningkids;
		int acstat;
		struct stat sbuf;
		time_t t;
		unsigned int sleeptime;

		signal(SIGCHLD, childsig);
		runningkids = 0;
		t = time(NULL);
		while ((pw = getpwent()) != NULL) {
			if (!pw->pw_dir || *pw->pw_dir != '/')
				continue;
			acstat = 0;
			/*
			 * Avoid unnecessary forks. The calendar file is only
			 * opened as the user later; if it can't be opened,
			 * it's no big deal. Also, get to correct directory.
			 * Note that in an NFS environment root may get EACCES
			 * on a chdir(), in which case we have to fork. As long
			 * as we can chdir() we can stat(), unless the user is
			 * modifying permissions while this is running.
			 */
			if (chdir(pw->pw_dir) ||
			    chdir(calendarHome)) {
				if (errno == EACCES) {
					acstat = 1;
					goto must_fork;
				}
				continue;
			}
			if (stat(calendarNoMail, &sbuf) == 0 ||
			    stat(calendarFile, &sbuf) != 0 ||
			    !S_ISREG(sbuf.st_mode))
				continue;
 must_fork:
			sleeptime = USERTIMEOUT;
			switch ((kid = fork())) {
			case -1:	/* error */
				warn("fork");
				continue;
			case 0:	/* child */
				(void)setlocale(LC_ALL, "");
				if (userswitch(pw))
					err(1, "unable to set user context (uid %u)",
					    pw->pw_uid);
				if (acstat) {
					calendarBaseDir = strdup(pw->pw_dir);
					if (!calendarBaseDir)
						err(1, NULL);
				}
				endpwent();
				if (acstat) {
					if (chdir(calendarBaseDir) ||
					    chdir(calendarHome) ||
					    stat(calendarNoMail, &sbuf) == 0 ||
					    stat(calendarFile, &sbuf) != 0 ||
					    !S_ISREG(sbuf.st_mode))
						exit(0);
				}
				cal();
				exit(0);
			}
			/* parent: wait a reasonable time, then kill child if
			 * necessary.
			 */
			runningkids++;
			kidreaped = 0;
			do {
				sleeptime = sleep(sleeptime);
				/* Note that there is the possibility, if the sleep
				 * stops early due to some other signal, of the child
				 * terminating and not getting detected during the next
				 * sleep.  In that unlikely worst case, we just sleep
				 * too long for that user.
				 */
				for (;;) {
					deadkid = waitpid(-1, &kidstat, WNOHANG);
					if (deadkid <= 0)
						break;
					runningkids--;
					if (deadkid == kid) {
						kidreaped = 1;
						sleeptime = 0;
					}
				}
			} while (sleeptime);

			if (!kidreaped) {
				/* It doesn't _really_ matter if the kill fails, e.g.
				 * if there's only a zombie now.
				 */
				(void)kill(kid, SIGTERM);
				warnx("uid %u did not finish in time", pw->pw_uid);
			}
			if (time(NULL) - t >= 3600*12)
				errx(2, "'calendar -a' took more than half a day; stopped at uid %u",
				    pw->pw_uid);
		}
		endpwent();
		for (;;) {
			deadkid = waitpid(-1, &kidstat, WNOHANG);
			if (deadkid <= 0)
				break;
			runningkids--;
		}
		if (runningkids)
			warnx("%d child processes still running when 'calendar -a' finished",
			    runningkids);
	} else {
		if ((calendarBaseDir = getenv("CALENDAR_DIR")) == NULL ||
		    !*calendarBaseDir)
			calendarBaseDir = ".";
		else if (chdir(calendarBaseDir))
			err(1, "%s: chdir", calendarBaseDir);
		cal();
	}

	exit(0);
}


void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: calendar [-abP] [-A num] [-B num] [-f calendarfile] "
	    "[-t [[[cc]yy]mm]dd]\n");
	exit(1);
}


void
childsig(int signo __unused)
{
}


static void
addcppargv(const char *fn, int ch)
{
	char *what;
	size_t len;
	static size_t ncppargv = 0;

	if (ncppargv == NCPPARGV)
		errx(1, "too many cpp(1) flags");
	switch (ch) {
	case 0:
		/* append sentinel nil */
		what = NULL;
		break;
	case 1:
		/* escape calendarFile */
		if (!(fn[0] == '/' || (fn[0] == '-' && !fn[1]))) {
			/* if it is not absolute / stdin */
			len = strlen(fn) + 1U;
			if (!(what = malloc(2U + len)))
				err(1, "malloc");
			what[0] = '.';
			what[1] = '/';
			memcpy(what + 2, fn, len);
			break;
		}
		/* if it is absolute / stdin, there is no need to escape */
		/* FALLTHROUGH */
	case 2: {
		/* direct assign ignoring const */
		union {
			char *cp;
			const char *ccp;
		} p;

		p.ccp = fn;
		what = p.cp;
		break;
	}
	default:
		/* pass on flag */
		len = strlen(fn) + 1U;
		if (!(what = malloc(2U + len)))
			err(1, "malloc");
		what[0] = '-';
		what[1] = ch;
		memcpy(what + 2, fn, len);
		break;
	}
	cppargv[ncppargv++] = what;
}
